### ISSUE TITLE
Debugger toggle not shown on virtualenv-based kernel

### ISSUE THREAD

**Issue by verdimrc** [jupyterlab/jupyterlab#11136]
## Description

The debugger toggle is shown only when an virtualenv has executed `python -m ipykernel install --user --name=kernel-name`.

However, the debugger toggle is not shown when the same virtualenv is automatically discovered (and registered as a kernel) by [environment_kernels](https://github.com/Cadair/jupyter_environment_kernels).

## Reproduce

1. Install [environment_kernels](https://github.com/Cadair/jupyter_environment_kernels)
2. Create a new virtualenv, (let's say `abc`)
3. Execute `/PATH/TO/VIRTUALENV/bin/python -m ipykernel install --user --name=my-kernel-abc`
2. Start jupyter lab
3. Open a notebook
    1. If choosing the manually registered kernel, then `my-kernel-abc`, and debugging toggle is shown.
    2. However, if choosing the automatically registered kernel, `virtualenv-my-kernel-abc`, and debugging toggle is not shown. 

## Expected behavior

Debugger toggle is shown on virtualenv-based kernel.

## Context

- Operating System and version: macOS Big Sur 11.6
- Browser and version: 93.0.4577.82 (Official Build) (x86_64)
- JupyterLab version: 3.1.13

```text
debugpy             1.4.3
ipykernel           6.4.1
ipython             7.27.0
jupyter-client      7.0.3
jupyter-core        4.8.1
jupyter-server      1.11.0
jupyterlab          3.1.13
jupyterlab-server   2.8.1
```

<details><summary>Troubleshoot Output</summary>
<pre>
pip list:
	Package             Version
	------------------- ---------
	anyio               3.3.1
	appnope             0.1.2
	argon2-cffi         20.1.0
	async-generator     1.10
	attrs               21.2.0
	Babel               2.9.1
	backcall            0.2.0
	bleach              4.1.0
	bokeh               2.4.0
	certifi             2021.5.30
	cffi                1.14.6
	chardet             4.0.0
	charset-normalizer  2.0.6
	debugpy             1.4.3
	decorator           5.1.0
	defusedxml          0.7.1
	entrypoints         0.3
	environment-kernels 1.1.1
	idna                3.2
	ipykernel           6.4.1
	ipython             7.27.0
	ipython-genutils    0.2.0
	ipywidgets          7.6.5
	jedi                0.18.0
	Jinja2              3.0.1
	json5               0.9.6
	jsonschema          3.2.0
	jupyter-bokeh       3.0.2
	jupyter-client      7.0.3
	jupyter-core        4.8.1
	jupyter-server      1.11.0
	jupyterlab          3.1.13
	jupyterlab-pygments 0.1.2
	jupyterlab-server   2.8.1
	jupyterlab-widgets  1.0.2
	MarkupSafe          2.0.1
	matplotlib-inline   0.1.3
	mistune             0.8.4
	nbclassic           0.3.2
	nbclient            0.5.4
	nbconvert           6.1.0
	nbformat            5.1.3
	nest-asyncio        1.5.1
	notebook            6.4.4
	numpy               1.21.2
	packaging           21.0
	pandocfilters       1.5.0
	parso               0.8.2
	pexpect             4.8.0
	pickleshare         0.7.5
	Pillow              8.3.2
	pip                 21.2.4
	prometheus-client   0.11.0
	prompt-toolkit      3.0.20
	ptyprocess          0.7.0
	pycparser           2.20
	Pygments            2.10.0
	pyparsing           2.4.7
	pyrsistent          0.18.0
	python-dateutil     2.8.2
	pytz                2021.1
	PyYAML              5.4.1
	pyzmq               22.3.0
	requests            2.26.0
	requests-unixsocket 0.2.0
	Send2Trash          1.8.0
	setuptools          57.4.0
	six                 1.16.0
	sniffio             1.2.0
	terminado           0.12.1
	testpath            0.5.0
	tornado             6.1
	traitlets           5.1.0
	typing-extensions   3.10.0.2
	urllib3             1.26.7
	wcwidth             0.2.5
	webencodings        0.5.1
	websocket-client    1.2.1
	widgetsnbextension  3.5.1
</pre>
</details>

<details><summary>Command Line Output</summary>

NOTE: the `virtualenv_yyyy` are the kernels automatically discovered by `environment_kernels`.

<pre>
[I 2021-09-24 09:56:28.041 ServerApp] jupyterlab | extension was successfully linked.
[W 2021-09-24 09:56:28.045 NotebookApp] 'kernel_spec_manager_class' has moved from NotebookApp to ServerApp. This config will be passed to ServerApp. Be sure to update your config before our next release.
[I 2021-09-24 09:56:28.398 ServerApp] nbclassic | extension was successfully linked.
[I 2021-09-24 09:56:28.401 ServerApp] Using EnvironmentKernelSpecManager...
[I 2021-09-24 09:56:28.402 ServerApp] Started periodic updates of the kernel list (every 3 minutes).
[I 2021-09-24 09:56:28.455 ServerApp] nbclassic | extension was successfully loaded.
[I 2021-09-24 09:56:28.456 LabApp] JupyterLab extension loaded from /Users/xxxx/.pyenv/versions/miniconda3-latest/envs/basedev/lib/python3.9/site-packages/jupyterlab
[I 2021-09-24 09:56:28.456 LabApp] JupyterLab application directory is /Users/xxxx/.pyenv/versions/miniconda3-latest/envs/basedev/share/jupyter/lab
[I 2021-09-24 09:56:28.460 ServerApp] jupyterlab | extension was successfully loaded.
[I 2021-09-24 09:56:28.461 ServerApp] Serving notebooks from local directory: /Users/xxxx/Desktop/abalone_build_train_deploy
[I 2021-09-24 09:56:28.461 ServerApp] Jupyter Server 1.11.0 is running at:
[I 2021-09-24 09:56:28.461 ServerApp] http://localhost:8888/lab?token=64024c5bcaa8e4714eb97f78e40759faca8f91ae5bde7c55
[I 2021-09-24 09:56:28.461 ServerApp]  or http://127.0.0.1:8888/lab?token=64024c5bcaa8e4714eb97f78e40759faca8f91ae5bde7c55
[I 2021-09-24 09:56:28.461 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
[C 2021-09-24 09:56:28.469 ServerApp] 
    
    To access the server, open this file in a browser:
        file:///Users/xxxx/Library/Jupyter/runtime/jpserver-69140-open.html
    Or copy and paste one of these URLs:
        http://localhost:8888/lab?token=64024c5bcaa8e4714eb97f78e40759faca8f91ae5bde7c55
     or http://127.0.0.1:8888/lab?token=64024c5bcaa8e4714eb97f78e40759faca8f91ae5bde7c55
[I 2021-09-24 09:56:28.474 ServerApp] Starting initial scan of virtual environments...
[I 2021-09-24 09:56:28.732 ServerApp] Found new kernels in environments: virtualenv_sm, virtualenv_basedev, virtualenv_chiller, virtualenv_sm-sdk, virtualenv_meta-gluonts
[W 2021-09-24 09:56:30.444 ServerApp] Kernelspec name virtualenv_sm cannot be found!
[W 2021-09-24 09:56:30.444 ServerApp] Kernelspec name virtualenv_basedev cannot be found!
[W 2021-09-24 09:56:30.445 ServerApp] Kernelspec name virtualenv_chiller cannot be found!
[W 2021-09-24 09:56:30.445 ServerApp] Kernelspec name virtualenv_sm-sdk cannot be found!
[W 2021-09-24 09:56:30.445 ServerApp] Kernelspec name virtualenv_meta-gluonts cannot be found!
[W 2021-09-24 09:56:30.746 ServerApp] Kernelspec name virtualenv_sm cannot be found!
[W 2021-09-24 09:56:30.746 ServerApp] Kernelspec name virtualenv_basedev cannot be found!
[W 2021-09-24 09:56:30.747 ServerApp] Kernelspec name virtualenv_chiller cannot be found!
[W 2021-09-24 09:56:30.747 ServerApp] Kernelspec name virtualenv_sm-sdk cannot be found!
[W 2021-09-24 09:56:30.747 ServerApp] Kernelspec name virtualenv_meta-gluonts cannot be found!
[I 2021-09-24 09:56:32.366 LabApp] Build is up to date
</pre>
</details>

<details><summary>Browser Output</summary>

<pre>
Paste the output from your browser Javascript console here, if applicable.

</pre>
</details>

**Comment by krassowski** [comment]
This looks like an issue with https://github.com/Cadair/jupyter_environment_kernels, not JupyterLab. The debugger support is detected based on specification of the kernel (kernelspec). You can list your kernels with `jupyter kernelspec list`; if the `kernel.json` has entry:

```
 "metadata": {
  "debugger": true
 }
```

the debugger will be enabled. `jupyter_environment_kernels` creates kernespecs in Python, and the [`convert_to_env_data()`](https://github.com/Cadair/jupyter_environment_kernels/blob/3da304550b511bda7d5d39280379b5ca39bb31bc/environment_kernels/envs_common.py#L24) function does not seem to add debugger information (it was written before debugger support arrived). I imagine that [`validate_IPykernel`](https://github.com/Cadair/jupyter_environment_kernels/blob/3da304550b511bda7d5d39280379b5ca39bb31bc/environment_kernels/envs_common.py#L63) and friends could detect the debugger support. In any case, without changing the contract it seems that this issue should be addressed in `jupyter_environment_kernels`; if change of the API is absolutely required (though I don't think so), then it should be discussed in other relevant jupyter repositories (possibly `jupyter_client`?).

Would you mind opening an issue over at https://github.com/Cadair/jupyter_environment_kernels?

**Comment by paloha** [comment]
I had the same problem. The debugger icon would not show up with ipykernel.
My versions: Ubuntu 20.04, Python 3.8.10, pip 21.2.4, jupyterlab 3.1.12, ipython 7.27.0, ipykernel 6.4.1.

I am creating my virtual environment using:
```
virtualenv .venv
source .venv/bin/activate
pip install jupyterlab # and other stuff
jupyter-lab
```
Now I do not see any debugger icon. 

I added the following code into the `kernel.json` as suggested by @krassowski:
```
 "metadata": {
  "debugger": true
 }
```
I have restarted the virtualenv and jupyterlab and now I see the debugger! :confetti_ball:  Thx @krassowski.
